### PR TITLE
canal: support ipv6 address

### PR DESCRIPTION
**Pull request by sinomoe**
canal: 
* address format support ipv6 with port like `[::1]:3306` in canal
* parse ipv6 address with optional port like `::1` or `[::1]:3306` in dumper

### CODE REVIEW

File: canal/canal.go [hunk 1]
```diff
@@ -481,17 +481,25 @@ func (c *Canal) prepareSyncer() error {
 	if strings.Contains(c.cfg.Addr, "/") {
 		cfg.Host = c.cfg.Addr
 	} else {
-		seps := strings.Split(c.cfg.Addr, ":")
```
Comment (lance6716) [inline canal/canal.go:484]: I think we can use https://pkg.go.dev/net#SplitHostPort

File: canal/canal.go [hunk 2]
```diff
@@ -481,17 +481,25 @@ func (c *Canal) prepareSyncer() error {
...
-			return errors.Errorf("invalid mysql addr format %s, must host:port", c.cfg.Addr)
+		ipv6 := strings.Count(c.cfg.Addr, ":") > 1
+		if ipv6 && !strings.ContainsAny(c.cfg.Addr, "[]") {
+			return errors.Errorf("invalid mysql ipv6 addr format %s, must [host]:port", c.cfg.Addr)
 		}
-
-		port, err := strconv.ParseUint(seps[1], 10, 16)
+		lastSep := strings.LastIndex(c.cfg.Addr, ":")
+		if !ipv6 && lastSep == -1 {
+			return errors.Errorf("invalid mysql ipv4 addr format %s, must host:port", c.cfg.Addr)
```
Comment (dveeden) [inline canal/canal.go:490]:
```suggestion
			return errors.Errorf("invalid MySQL IPv4 addr format %s, must host:port", c.cfg.Addr)
```

File: canal/canal.go [hunk 3]
```diff
@@ -481,17 +481,25 @@ func (c *Canal) prepareSyncer() error {
...
+		}
+		addr := strings.Trim(c.cfg.Addr[:lastSep], "[]")
+		ip := net.ParseIP(addr)
+		if ip == nil {
+			return errors.Errorf("invalid mysql ip format %s", addr)
```
Comment (dveeden) [inline canal/canal.go:495]:
```suggestion
			return errors.Errorf("invalid MySQL IP-address format %s", addr)
```

**Review by dveeden — COMMENTED** [review]
Should we try to support things like `:3307` (e.g. only a port, not a host?)

File: canal/canal.go (hunk 1 above)
Comment (sinomoe) [reply canal/canal.go:484]: Good idea, let me check this

File: canal/canal.go [hunk 4]
```diff
@@ -481,17 +481,25 @@ func (c *Canal) prepareSyncer() error {
 	if strings.Contains(c.cfg.Addr, "/") {
 		cfg.Host = c.cfg.Addr
 	} else {
-		seps := strings.Split(c.cfg.Addr, ":")
-		if len(seps) != 2 {
-			return errors.Errorf("invalid mysql addr format %s, must host:port", c.cfg.Addr)
+		ipv6 := strings.Count(c.cfg.Addr, ":") > 1
+		if ipv6 && !strings.ContainsAny(c.cfg.Addr, "[]") {
+			return errors.Errorf("invalid mysql ipv6 addr format %s, must [host]:port", c.cfg.Addr)
```
Comment (dveeden) [inline canal/canal.go:486]:
```suggestion
			return errors.Errorf("invalid MySQL IPv6 address format %s, must [host]:port", c.cfg.Addr)
```

File: canal/canal.go (hunk 2 above)
Comment (sinomoe) [reply canal/canal.go]: done

File: mysql/const.go [hunk 5]
```diff
@@ -172,6 +172,7 @@ const (
 
 const (
 	DEFAULT_ADDR                  = "127.0.0.1:3306"
+	DEFAULT_IPv6_ADDR             = "[::1]:3306"
```
Comment (dveeden) [inline mysql/const.go:175]:
```suggestion
	DEFAULT_IPV4_ADDR                  = "127.0.0.1:3306"
	DEFAULT_IPV6_ADDR             = "[::1]:3306"
```

Not sure if we should uppercase the `V` here or not..

Comment (sinomoe) [reply mysql/const.go]: Changing `DEFAULT_ADDR` to `DEFAULT_IPv6_ADDR` might break compatibility, maybe it's better to leave it unchanged.

**Comment by sinomoe** [comment]
> What about tests and CI? Anything we can do to test this and/or make it easier to test?

Test cases and CI have been completed.

File: dump/dumper.go [hunk 6]
```diff
@@ -202,10 +202,25 @@ func (d *Dumper) Dump(w io.Writer) error {
 	if strings.Contains(d.Addr, "/") {
 		args = append(args, fmt.Sprintf("--socket=%s", d.Addr))
 	} else {
-		seps := strings.SplitN(d.Addr, ":", 2)
-		args = append(args, fmt.Sprintf("--host=%s", seps[0]))
-		if len(seps) > 1 {
-			args = append(args, fmt.Sprintf("--port=%s", seps[1]))
+		ipv6 := strings.Count(d.Addr, ":") > 1
```
Comment (lance6716) [inline dump/dumper.go:205]: this part can also use `net.SplitHostPort`, just like canal/canal.go

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: dump/dumper.go (hunk 6 above)
Comment (sinomoe) [reply dump/dumper.go]: done

**Review by lance6716 — APPROVED** [review]
ptal @dveeden

**Review by dveeden — APPROVED** [review]

**Pull request merged**